solidity/src/Polymer7683.sol

```diff
@@ -0,0 +1,221 @@
+        // 1. Verify the destination contract is registered
+        address expectedEmitter = destinationContracts[destinationChainId];
+        if (expectedEmitter == address(0)) revert UnregisteredDestinationChain();
+
+        // 2. Verify event using Polymer prover
+        (
+            uint32 chainId,
+            address actualEmitter,
+            bytes memory topics,
+            bytes memory data
+        ) = prover.validateEvent(eventProof);
+
+        // 3. Validate the chain ID matches our expected destination chain
+        if (chainId != destinationChainId) {
+            revert InvalidChainId();
+        }
+
+        // 4. Validate the emitter matches our registered destination contract
+        if (actualEmitter != expectedEmitter) revert InvalidEmitter();
+
+        // 4. Verify this event hasn't been processed before (prevent replay)
+        bytes32 eventHash = keccak256(abi.encodePacked(eventProof, logIndex, destinationChainId));
+        if (processedEvents[eventHash]) revert EventAlreadyProcessed();
+        processedEvents[eventHash] = true;
+
+        // 5. Parse the BatchOrdersFilled event data (with empty fillerData for refunds)
+        (bytes32[] memory eventOrderIds, bytes[] memory ordersFillerData) = 
+            abi.decode(data, (bytes32[], bytes[]));
+
+        // 6. Verify that this is actually a refund event which is indicated by the empty filler data.
+        if (ordersFillerData.length != 0) revert InvalidEventData();
+
+        // 7. Validate order IDs match what was requested
+        if (orderIds.length != eventOrderIds.length) revert InvalidEventData();
+        for (uint256 i = 0; i < orderIds.length; i++) {
+            if (orderIds[i] != eventOrderIds[i]) revert InvalidEventData();
+        
+            // 8. Process refund for each order
+            _handleRefundOrder(orderIds[i]);
+        }
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Dispatches a settlement instruction by emitting an event that will be proven on the origin chain
+     * @param _originDomain The domain to which the settlement message is sent
+     * @param _orderIds The IDs of the orders to settle
+     * @param _ordersFillerData The filler data for the orders
+     */
+    function _dispatchSettle(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds,
+        bytes[] memory _ordersFillerData
+    ) internal override {
+        emit BatchOrdersFilled(localChainId, _orderIds, _ordersFillerData);
+    }
+
+    /**
+     * @notice Dispatches a refund instruction by emitting an event that will be proven on the origin chain
+     * @param _originDomain The domain to which the refund message is sent
+     * @param _orderIds The IDs of the orders to refund
+     */
+    function _dispatchRefund(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds
+    ) internal override {
+        emit BatchOrdersFilled(localChainId, _orderIds, new bytes[](0));
```

Comment:
Shouldn't this be a different event?. 
If I don't get this wrong, having the same event on both settle and refund would allow you to call `handleRefundWithProof` or `handleSettlementWithProof` regardless of where the event was emitted.

Reply:
the original design was a bit hacky - I've refactored to use the events from the base contract